### PR TITLE
chore: Prettier pass

### PR DESCRIPTION
**Pull request by agg23**
### User facing changelog
None

### Additional details
Switches the monorepo's style linter to Prettier, away from custom ESLint configurations. We specifically use `prettierx` plugins to disable the `indentChains` rule (see comments below).

For easy reviewing, commits up until (inclusive) eafeaa9201b4a1f23104495136222be469fc672c do not have any of the actual linting changes, only the configuration. The commits starting at 1dcebf339ae3710bb4b841c9b6d9b4b3ce439610 are test fixes and merges.

#### Prettier
- Pros
  - Commonly used in the JS and TS worlds. Is beginning to define **the** way JS is formatted
  - Very consistent formatting. Rarely does inconsequential changes in the input (spacing, newlines, etc...) change the output
  - Few options for bikeshedding or configuration; little to go wrong. We would be able to drastically reduce the number of specially configured rules in our ESLint configs across the repo
  - First class TypeScript and SASS support
  - Everything is auto-fixable
- Cons
  - Few options for configuration. If you don't like how it formats something, you probably can't change the config to produce different output. The best you can do is `/* prettier-disable */`
  - An additional build dependency. While we can (and will) integrate Prettier into ESLint, Prettier is still an extra package to be pulled in
  - As with any format changes, will introduce a large PR (like this one) to update _all_ of the formatting
  - We will still be using ESLint for non-format linting

#### ESLint
- Pros
  - Highly configurable. Many plugins to choose from
  - Already in use. Nothing new to do (except maybe find and implement more strict rules)
- Cons
  - Poor TS support (see Examples below)
  - Many rules (such as `max-len`) are not auto-fixable, requiring user intervention
  - Less consistent than Prettier (see Examples below)

### Examples

1. Written in several different forms, both flat and various multiline shapes.
  - ESLlint

    ```typescript
    export const Icon = createStory<{
      onPrefixClick: () => void
      onSuffixClick: () => void
        }
        >(() => {})
    ```
    
    This is the best autofix can do. It actually complains that there's a space between the function name and the arguments, completely ignoring the fact that TypeScript exists. It is unable to produce any formatting changes if the starting code is
    
    ```typescript
    export const Icon = createStory<{
      onPrefixClick: () => void, onSuffixClick: () => void}>(() => {})
    ```

    If the expression is written entirely in a single line, ESLint leaves it alone, which is hard to read in my opinion.

  - Prettier

    ```typescript
    export const Icon = createStory<{
      onPrefixClick: () => void;
      onSuffixClick: () => void;
    }>(() => {});
    ```

    Consistent spacing. Generic object literal maintains inner and outer indentation.

2. Originally written in a single line
  - ESLint

    ```typescript
      new Promise<number>((resolve) => {
        let input = 5
    
        input *= 25; resolve(input)
      }).then((v) => {
        console.log(`Received value ${v}`)
    
        return v % 50
      }).then((v) => `#${v}`).then((v) => {
        console.log(v); console.log('Stripping value')
    
        return v.replace('#', '')
      })
    ```

    ESLint is correctly able to split the output into multiple lines, but it produces some oddities. Why is `input *= 25; resolve(input)` all on one line? Why are the last two `.then`s stacked into a single line, instead of having a newline between them (this could probably be fixed with additional configuration)?

  - Prettier
    
    ```typescript
    new Promise<number>((resolve) => {
      let input = 5;
  
      input *= 25;
      resolve(input);
    })
      .then((v) => {
        console.log(`Received value ${v}`);
  
        return v % 50;
      })
      .then((v) => `#${v}`)
      .then((v) => {
        console.log(v);
        console.log("Stripping value");
  
        return v.replace("#", "");
      });
    ```

    Newlines are where _I_ expect them to be. Each call chain (`.then`) is stacked vertically on top of each other with newlines in-between.

3. Written in a single line
  - ESLint
  
    ```typescript
    type Expression = LongGenericNameWithManyCharacters<{foo: AnotherLongGenericName<WrapObjectWithYetAnotherLongGenericName<{innerLiteral: 'foo'}>>}>
    ```
    
    ESLint doesn't change anything about this expression. It's fine with it being a 147 char line with nested object literals.

  - Prettier

    ```typescript
    type Expression = LongGenericNameWithManyCharacters<{
      foo: AnotherLongGenericName<
        WrapObjectWithYetAnotherLongGenericName<{ innerLiteral: "foo" }>
      >;
    }>;
    ```

4. Written with the generics and function args inline
  - ESLint

    ```typescript
    const someFunc = <T extends LongGenericNameWithManyCharacters<{test: 'foo'}>, S extends AnotherLongGenericName<number>>(input: T): S => {
      return 4 as S
    }
    ```
    
    As with the last example, ESLint leaves the generics as is

  - Prettier

    ```typescript
    const someFunc = <
      T extends LongGenericNameWithManyCharacters<{ test: "foo" }>,
      S extends AnotherLongGenericName<number>
    >(
      input: T
    ): S => {
      return 4 as S;
    };
    ```
    
    The generics and function args are split into newlines

### CODE REVIEW

**Comment by brian-mann** [comment]
I'm on board with this so long as you can prevent prettier from doing two things:

### 1. do not nest `.then()` on a newline. 

The `.then()` should be flattened to the same indention as the outer promise. example:

```js
// good
new Promise((resolve) => {

})
.then(...)
.then(...)
```

```js
// bad (the way prettier does this)
new Promise((resolve) => {

})
  .then(...)
  .then(...)
```

This leaves dangling `)}` that don't match the indentation for really **no good reason**.

### 2. do not force a single chainable `.then()` to be on the same line

It should accept either a single line or new line

```js
// good
new Promise((resolve) => {

})
.then(...)
```

```js
// bad (the way prettier does this)
new Promise((resolve) => { ... }).then(...)
```

Basically prettier nesting and rules surrounding when to force things into a single line is always bad, and should be left to the developer to decide when to insert the newline.

### 3. do not force multiple chainables to be multiline

```js
// good
Promise.resolve().tap().catch()
```

```js
// also good
Promise
.resolve()
.tap()
.catch()
```

```js
// bad (the way prettier does this)
Promise.resolve()
  .tap()
  .catch()
```

In addition, it means adding a single chainable call forces the calls to be across multiple lines.

**Comment by agg23** [comment]
@brian-mann Yes, the plan is to use `prettier-eslint`. Part of the reasoning behind Prettier is to prevent bikeshedding, which we will get into if we start talking about promise chaining style. I happen to like how Prettier formats it 😄 

In general, what we're looking for is a way for everything to be consistent and require as little configuration as possible on our side, so stuff can't go wrong. As soon as we start looking for forks or modifications to Prettier, we're getting into the "stuff _can_ go wrong territory".

**Comment by lmiller1990** [comment]
I don't want to format code manually. Could we just bite the bullet and accept that `then` changes white space a little? Is this a big deal? I like the idea of opinionated defaults, whatever they happen to be.

**Comment by flotwig** [comment]
All the defaults look fine to me, *except*, as Brian says, the nesting of `.then`s (and I guess other chainables?)

To me, this:

```js
promise
.then(() => {
})
.then(() => {
})
```

is way more clear than this:

```js
promise
  .then(() => {
  })
  .then(() => {
  })
```

If that can be fixed, I think this PR looks good.

> Could we just bite the bullet and accept the defaults? I like the idea of opinionated defaults, whatever they happen to be. The whole point of prettier is bike shedding - hit save, run the linter, push and move on.

I don't think I understand this line of reasoning, why wouldn't we want to choose the best (most readable) linting options for our team? IMO this indentation change is clearly a readability setback, so let's take the time now, while we're talking about linting, to get it right.

**Comment by agg23** [comment]
> I don't think I understand this line of reasoning, why wouldn't we want to choose the best (most readable) linting options for our team? IMO this indentation change is clearly a readability setback, so let's take the time now, while we're talking about linting, to get it right.

The reasoning is because it prevents exactly this discussion. I find it more readable to be indented, as does @tgriesser and most (all?) of CT. When a tool tells you "this is what you must do", you don't have people arguing about stuff like this.

**Comment by lmiller1990** [comment]
If it's an easy change to make, sure, let's do it. I don't have any strong opinion on how the code is actually formatted, but I do have a strong belief that formatting/fixing linting errors by hand is not time well spent.

We could try using https://github.com/brodybits/prettier-plugin-x. It seems like it will just let us add the extra promise indentation rule without much. I am not convinced patch-package is a good long term solution - won't that make it difficult to upgrade prettier in the future?

**Comment by cowboy** [comment]
FWIW, at my last job, we standardized on prettier (using prettier-eslint), and it worked very well. Many of us had lists like @brian-mann does (myself included) but at the end of the day, we just decided to use stock prettier config and not worry about it. We did, however, set the built-in options to disable semicolons and use single quotes, because we were migrating from [JavaScript Standard Style](https://standardjs.com/rules.html) and wanted a little less noise in our cleanup commits.

**Comment by cowboy** [comment]
I want to add that, for me at least, it's very nice to be able to add a widely-used, well-supported plugin into my editor, do almost zero configuration, and know that my code is going to be formatted properly and consistently with the rest of our codebase!

**Comment by lmiller1990** [comment]
I'd like to see some movement on this. Can we just bite the bullet, so to speak, and go with whatever @agg23 has configured here?

**Comment by brian-mann** [comment]
For me this is beyond an internal style change for our team - this is about affecting and changing the very nature of how our product **appears** to users. That's why this conversation elevates outside of being just bikeshedding.

I am against merging this without the chaining formatting change because I am against the enforcement that our product test code and documentation conforms to this.

FWIW I posted links to plugins that remove this formatting change without forking prettier. It's literally as simple as dropping those into prettier and does not require a heavy lift whatsoever.

**Review by brian-mann — CHANGES_REQUESTED** [review]
See these comments: 
- https://github.com/cypress-io/cypress/pull/15788#issuecomment-818448019
- https://github.com/cypress-io/cypress/pull/15788#issuecomment-837154017

**Comment by agg23** [comment]
> For me this is beyond an internal style change for our team - this is about affecting and changing the very nature of how our product appears to users. That's why this conversation elevates outside of being just bikeshedding.
>
> I am against merging this without the chaining formatting change because I am against the enforcement that our product test code and documentation conforms to this. This changes our product, and it's not an internal only change.

@brian-mann I will make the change if you force me to, though I will argue that this is an improvement to docs/how Cypress appears (regardless about what you think about the newlines with chainables), due to one nice little coincidence:

The common indentation in the JS community (and used by Cypress) is two spaces. Extremely conveniently to us, our core interaction point is the `cy` prefix, which is also two characters. This results in the following formatting:
```typescript
cy.get(`${el}.inline_blank`)
  .click()
  .then(() => {})
```

Notice how the periods and the chainable methods are lined up perfectly.

**Comment by brian-mann** [comment]
For the record and for any future discussions see our internal slack discussion here:

https://cypressio.slack.com/archives/C01TQQCQLJY/p1620674637231400?thread_ts=1620672817.219600&cid=C01TQQCQLJY

For anyone else following along, the indentation rules of prettier are arbitrary, inconsistent, and mathematically (and geometrically) do not make sense when you consider indentation as a representation of lexical scope demonstrated by these examples:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEATOBnMAnAlgIzgAoByAMwghIBoACIgSloF4A+W4AHSltoHo+tXFHSw4qbr1wxSAOnk16TNh0m9+g4aPipaAQwDme4WoC+Dbue7d0WPIVIUqdRi3ZceGoSIQ610uQUXZXc1XgFvbXF9IxNPXjAAT1knRll8PWw0jIAvNJgACwQiVxUPdXUIgHcCvRghWiq9WDooOCraLAgABzgw2itPQcGbTBwCYnJKRVLQzwitX3F-GRJ5WRmQ1XivRbFdQ2MofqT+lMpGM4ysix30vTzbitlC4tntivDBGrqGppbaG0Ol1ev1BrxhrdRnYJo5psE3B9dj59itAhsEWV+gsUToYkd+gAFbAQAC2uAwcFk2EwEAANgA3SbrEgMF5FKAlLblT5Oa65S47cEDW4jHwwhxTZxKRE8nFRCSeAJrIIyrE7eVLA6xY47YlkilUmkYelM9Gsl56bpEMhQNmvTnvHkVPmZAVPdTCyHWcXjSVOTaytSa1FK1YszFzCohvGHOIVfXkylnY2m5kKD28dlvbn9Xiu7DusGZ2iyMB1MAFLlBna8YtmUUMEDUEA9GC4aAYZCgTIkqqEzIILsoPR0pqJLst-DYPRgADWcBgAGVurPhAZkDBsABXOAtuCkwiodCoAAyzQM28McAAYhBsKS6u2oBuR9uYBBmyACjBSXSAOoFNImCrmAcBLkO0i4Ay0iJMg4AYJOIDCJS2AwMShiPsgZCjpSLYAFYYAAHgAQjO86LkueiknAp7CHA2G4XuICEURS7rnScAAIrbhA8CMXSeEgKu2CofBGSEHSX7dHgsD-rgqCFMgAAcAAMLYyRAlL-jO3TwTJmBwNgTJfgAjrx8D6t0w4gHoGAALRtOI4hfjS5m4DSGEGFhSA4YJzGUuSm47gFHHcRZDG+UxLYwHo+DyYpBTIAATDFM64HS64AMJkj5ICYAArF+26UgAKnFw5+UJDK7gAkriS7jN0MAAIIiEuMCJJxAmUqYphAA

File: packages/driver/cypress/integration/commands/actions/scroll_spec.js [hunk 1]
```diff
@@ -396,153 +459,174 @@ describe('src/cy/commands/actions/scroll', () => {
...
-        })
-
-        return null
-      })
+    describe(
+      'errors',
+      {
+        defaultCommandTimeout: 50,
+      },
+      () => {
```
Comment (flotwig) [inline packages/driver/cypress/integration/commands/actions/scroll_spec.js:481]: another example of something not matching how we prescribe to users to do this. we'd suggest something more compact like:

```suggestion
    describe('errors', { defaultCommandTimeout: 50 }, () => {
```

**Review by flotwig — DISMISSED** [review]
I see a couple of smaller instances where the formatting doesn't match how we prescribe to users to use Cypress. I've left comments, however, I'm okay with merging it, we can iterate ("bike shed") over these details later on. Especially since AFAICT this change doesn't touch `@packages/example`.

File: packages/driver/cypress/integration/commands/actions/scroll_spec.js (hunk 1 above)
Comment (flotwig) [reply packages/driver/cypress/integration/commands/actions/scroll_spec.js:481]: yeah. i wonder if there could be a prettier-cypress-plugin someday. IMO, in tests, it makes sense to have some specialized linting, since the `describe(...it(...` stuff is almost like a DSL

Comment (brian-mann) [reply packages/driver/cypress/integration/commands/actions/scroll_spec.js:481]: Thanks @flotwig this is precisely why I was saying this is a full on product change, and something we need to be very prescriptive and careful about. The `describe` and `it` is a fluid DSL that's not superficial code - it's **the format** and outline for the contract between value and our users.

**Comment by brian-mann** [comment]
Can we look to see if `prettierx` has these options built in?

I think the formatting of `describe` and `it` over multiple lines is horrendous and has to be fixed. The other stuff I'm not as offending by but ideally these are the rules for our product:

```js
// good
cy.get('...') 

// bad
cy
.get('...') 

// good
cy.get('...').click()
cy.get('...').find(...).click()
cy.get('...').find(...).click().then(($a) => { 
})

// good
cy
.get('...')
.find(...')
.click()
.then(($a) => {

})

// bad
cy.get('...') 
.find('...')
.click()
.then(($a) => {

})
```

**Comment by brian-mann** [comment]
I’m still thinking about how to handle the exceptions. I need more time but have some ideas about a compromise. Do not merge yet.

File: cli/lib/cli.js [hunk 2]
```diff
@@ -395,18 +391,12 @@ module.exports = {
...
 
-    addCypressRunCommand(program)
-    .action((...fnArgs) => {
+    addCypressRunCommand(program).action((...fnArgs) => {
       debug('running Cypress with args %o', fnArgs)
-      require('./exec/run')
-      .start(parseVariableOpts(fnArgs, args))
-      .then(util.exit)
-      .catch(util.logErrorExit1)
+      require('./exec/run').start(parseVariableOpts(fnArgs, args)).then(util.exit).catch(util.logErrorExit1)
```
Comment (jennifer-shehane) [inline cli/lib/cli.js:399]: I feel like all  of these calls in the CLI have ended up a lot less clear to read fwiw.

File: cli/lib/errors.js [hunk 3]
```diff
@@ -228,10 +224,12 @@ const invalidTestingType = {
...
  * @param {'close'|'event'} eventName Child close event name
  * @param {string} signal Signal that closed the child process, like "SIGBUS"
-*/
+ */
 const childProcessKilled = (eventName, signal) => {
   return {
-    description: `The Test Runner unexpectedly exited via a ${chalk.cyan(eventName)} event with signal ${chalk.cyan(signal)}`,
+    description: `The Test Runner unexpectedly exited via a ${chalk.cyan(eventName)} event with signal ${chalk.cyan(
+      signal
+    )}`,
```
Comment (jennifer-shehane) [inline cli/lib/errors.js:232]: Why is this first call to chalk inline and the second one is spread multiline?

File: cli/test/lib/cli_spec.js [hunk 4]
```diff
@@ -102,11 +102,7 @@ describe('cli', () => {
     }
 
     const sanitizePlatform = (text) => {
-      return text
-      .split(os.eol)
-      .map(replacePlatform)
-      .map(replaceCypressVersion)
-      .join(os.eol)
+      return text.split(os.eol).map(replacePlatform).map(replaceCypressVersion).join(os.eol)
```
Comment (jennifer-shehane) [inline cli/test/lib/cli_spec.js:105]: also not a fan.

File: packages/reporter/cypress/integration/suites_spec.ts [hunk 5]
```diff
@@ -31,99 +31,80 @@ describe('suites', () => {
...
   describe('expand and collapse', () => {
     it('is expanded by default', () => {
       cy.contains('suite 1')
-      .parents('.collapsible').as('suiteWrapper')
+      .parents('.collapsible')
+      .as('suiteWrapper')
       .should('have.class', 'is-open')
-      .find('.collapsible-content').eq(0)
+      .find('.collapsible-content')
+      .eq(0)
```
Comment (jennifer-shehane) [inline packages/reporter/cypress/integration/suites_spec.ts:54]: I don't understand why one test allows inline and the other requires indention. Is it reaching a max length? Sorry, I'm completely ignorant to prettier.

**Review by jennifer-shehane — COMMENTED** [review]
I agree with Brian's points, especially concerning Cypress tests in our code. Personally, I'm not a fan of forcing chains to be inline if it is more readable to a human being on separate lines and vice versa.  

I don't see the harm in having a highly customized linting configuration if that ends up being the best solution for our codebase.

File: cli/lib/cli.js (hunk 2 above)
Comment (agg23) [reply cli/lib/cli.js:399]: I agree. This is because of the changes Brian requested.

File: cli/lib/errors.js (hunk 3 above)
Comment (agg23) [reply cli/lib/errors.js:232]: Line got too long

**Comment by jennifer-shehane** [comment]
Converting this PR to draft to quiet the notifications on it.

**Comment by tgriesser** [comment]
Closing for now - we can revisit in the future.
